### DISCUSSION TITLE
Stripe code refactoring options

### DISCUSSION THREAD

**Discussion by Shriukan33** [transport-nantes/tn_web#1121] [General]
Stripe app has a clunky way to work right now. 

The philosophy was to mimic greenpeace's donation form, which is a 3 step form :
* 1st step is amount and recurrence
* 2nd step is personal info
* 3rd step is payment

because GP's form doesn't reload the page between steps, I tried to fit 2 forms into 1 : I had to validate the 1st step and then the 2nd.
For that purpose I used [reportValidity](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/reportValidity) to validate a form solely on HTML basis and then display a different part of it. 

```js
var step_1_completed = false
var step_2_completed = false
// Hides part 1 and displays part 2.
var valid_amount_form = false
document.forms['amount_form'].addEventListener('submit', function() {
    valid_amount_form = document.forms['amount_form'].reportValidity();
    if (valid_amount_form == true) {
        document.getElementById("amount_form").style.display = "none"
        document.getElementById("donation_form").style.display = "block"
        document.getElementById("progress_bar_step_2").classList.add("active")
        step_1_completed = true
    }
  }, false);
```

I don't think this is the correct approach anymore. This forces us to partially validate a form, and only according to HTML's criterion : The field is field, with an appropriate type for value, and it makes a complicated javascript logic.

I think that the two first steps should actually be 2 different forms that are POSTed to Django and validated with a 200 response, possibly with a JSONResponse that holds the relevant information.

To keep the Single Page Application feel we were looking for, we could use either [htmx](https://htmx.org/) or edit the dom to embed a new form directly, rendered by django using ajax callback. 

The flow would be :
* User fills a form that has only the payment / subscription part and clicks continue
* Ajax sends the form to django which in turn validates it and returns either the new form (personal information) as a string that replaces the former form on the front-end, or the submitted form and its errors if the form didn't go through validation.
* The user fills the new form and hit continue again.
* Same logic except that we can return a JSONresponse with the session ID in case of success and redirect to Stripe's session once the response received.

Now another approach without the "no reload" touch, we could simply POST the 1st step, validate, and add needed info into GET args and load the 2nd step's form. We'd validate that the proper GET args are there to make sure user went through step 1, and then simply make an AJAX that would redirect on a 200 once step 2 is submitted and valid.

Maybe there are other ways, but these seem more reasonable than the current convoluted JS we have.

**Comment by JeffAbrahamson** [comment]
The SPA isn't a requirement.  If it complicates life, we should drop it.  The only important bit is that the transitions are fast enough and can't fail in ways that the current app doesn't.

Note that forms may have [initial values](https://docs.djangoproject.com/en/4.1/topics/forms/modelforms/#providing-initial-values), which is a convenient way to give the appearance of progressively filling out a form.  The initial values may be hidden in the second form to increase the illusion of state.

**Reply by Shriukan33** [reply]
Yes, adding hidden fields init with first form's values in the 2nd form would probably work just as well.
